### PR TITLE
add a simple Snakemake profile for Slurm

### PR DESCRIPTION
**Pull request by xinyuejohn**
Add a simple Snakemake profile for Slurm.

Adapted from https://github.com/jdblischak/smk-simple-slurm.

### CODE REVIEW

**Comment by smilesun** [comment]
@xinyuejohn , i am going to resolve the conflict now

**Comment by smilesun** [comment]
as agreed, this pr will be replace to a new pr, so i close it here

**Comment by smilesun** [comment]
PyTorch CUDA memory summary, device ID 0                 |
|---------------------------------------------------------------------------|
|            CUDA OOMs: 0            |        cudaMalloc retries: 0         |
|===========================================================================|
|        Metric         | Cur Usage  | Peak Usage | Tot Alloc  | Tot Freed  |
|---------------------------------------------------------------------------|
| Allocated memory      |       0 B  |    3226 MB |   21471 GB |   21471 GB |
|       from large pool |       0 B  |    3221 MB |   21455 GB |   21455 GB |
|       from small pool |       0 B  |      10 MB |      16 GB |      16 GB |
|---------------------------------------------------------------------------|
| Active memory         |       0 B  |    3226 MB |   21471 GB |   21471 GB |
|       from large pool |       0 B  |    3221 MB |   21455 GB |   21455 GB |
|       from small pool |       0 B  |      10 MB |      16 GB |      16 GB |
|---------------------------------------------------------------------------|
| GPU reserved memory   |     980 MB |    3480 MB |    3480 MB |    2500 MB |
|       from large pool |     976 MB |    3468 MB |    3468 MB |    2492 MB |
|       from small pool |       4 MB |      12 MB |      12 MB |       8 MB |
|---------------------------------------------------------------------------|
| Non-releasable memory |       0 B  |  423934 KB |    6454 GB |    6454 GB |
|       from large pool |       0 B  |  421328 KB |    6432 GB |    6432 GB |
|       from small pool |       0 B  |    5193 KB |      21 GB |      21 GB |
|---------------------------------------------------------------------------|
| Allocations           |       0    |    1036    |    1591 K  |    1591 K  |
|       from large pool |       0    |     111    |     721 K  |     721 K  |
|       from small pool |       0    |    1004    |     869 K  |     869 K  |
|---------------------------------------------------------------------------|
| Active allocs         |       0    |    1036    |    1591 K  |    1591 K  |
|       from large pool |       0    |     111    |     721 K  |     721 K  |
|       from small pool |       0    |    1004    |     869 K  |     869 K  |
|---------------------------------------------------------------------------|
| GPU reserved segments |      13    |      69    |      69    |      56    |
|       from large pool |      11    |      63    |      63    |      52    |
|       from small pool |       2    |       6    |       6    |       4    |
|---------------------------------------------------------------------------|
| Non-releasable allocs |       0    |     115    |     815 K  |     815 K  |

**Comment by smilesun** [comment]
/var/spool/slurmd/job12367024/slurm_script: line 3: 224719 Killed                  DOMAINLAB_CUDA_START_SEED='' /home/aih/xudong.sun/anaconda3/bin/python -m snakemake --snakefile '/home/aih/xudong.sun/DomainLab/domainlab/exp_protocol/benchmark.smk' --target-jobs 'run_experiment:index=5' --allowed-rules 'run_experiment' --cores 'all' --attempt 4 --force-use-threads --resources 'mem_mb=1000' 'mem_mib=954' 'disk_mb=1000' 'disk_mib=954' --wait-for-files '/home/aih/xudong.sun/DomainLab/.snakemake/tmp.g36699hz' 'zoutput/benchmarks/test_bloodcell_benchmark/hyperparameters.csv' --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --keep-incomplete --rerun-triggers 'params' 'mtime' 'input' 'code' 'software-env' --skip-script-cleanup --use-conda --conda-frontend 'mamba' --conda-base-path '/home/aih/xudong.sun/anaconda3' --wrapper-prefix 'https://github.com/snakemake/snakemake-wrappers/raw/' --configfiles '/home/aih/xudong.sun/DomainLab/examples/yaml/test_helm_benchmark.yaml' --printshellcmds --latency-wait 60 --scheduler 'greedy' --scheduler-solver-path '/home/aih/xudong.sun/anaconda3/bin' --default-resources 'mem_mb=1000' 'disk_mb=max(2*input.size_mb, 1000)' 'tmpdir=system_tmpdir' 'partition=gpu_p' 'qos=gpu' --mode 2
slurmstepd: error: Detected 1 oom-kill event(s) in StepId=12367024.batch cgroup. Some of your processes may have been killed by the cgroup out-of-memory handler.

**Comment by smilesun** [comment]
@xinyuejohn @rahulbshrestha , which configuration did you use you have't met this error? 

what is the diffrent between mem_mb=1000 and --mem=60G? 

I changed mem_mb=10000 and will see the results

**Comment by rahulbshrestha** [comment]
@smilesun this PR is a bit confusing, are the files such as "hyperparameter_sampling_error.py" supposed to be included or not? Also, did running this branch work for you as compare to the branch in https://github.com/marrlab/DomainLab/pull/146?

**Comment by smilesun** [comment]
> @smilesun this PR is a bit confusing, are the files such as "hyperparameter_sampling_error.py" supposed to be included or not? Also, did running this branch work for you as compare to the branch in #146?
@rahulbshrestha 
Hi,
0. branch #146 has missing files as I wrote in the PR.  But you can resolve the issues I posted in that PR to update the files in this PR. 
1. hyper**_error should be removed, I am just curious to see what is inside there 
2. it works on my laptop by the following

```
rm zoutput # first clean up the output directory
git checkout snakemake_slurm
./examples/benchmark/run_snakemake.sh  

```

the last error is what I created deliberately,  for issue 
https://github.com/marrlab/DomainLab/issues/152
where on my laptop the rule "agg_partial_results" does not work. 


Did you met any problem on your laptop? Could you post the error message in an issue?

**Comment by smilesun** [comment]
@rahulbshrestha , sorry I think I misunderstood, I was trying to resolve the other issue which can be reproduced on laptop, i guess we are now talking the helm cluster? Did you meet any problem there? What is the difference of the current  version to the version you have used for generating the results for the presentation?

**Comment by rahulbshrestha** [comment]
> @rahulbshrestha , sorry I think I misunderstood, I was trying to resolve the other issue which can be reproduced on laptop, i guess we are now talking the helm cluster? Did you meet any problem there? What is the difference of the current version to the version you have used for generating the results for the presentation?

Hey @smilesun , nope there was no issue in the Helmholtz cluster, we ran the latest changes from this branch.

File: examples/yaml/slurm/config.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    --gres=gpu:1
+    --nice=10000
+    -c 2
+    --mem=60G
+    --job-name=smk-{rule}-{wildcards}
+    --output=logs/{rule}/{rule}-{wildcards}-%j.out
+default-resources:
+  - partition=gpu_p
+  - qos=gpu
+  - mem_mb=60000
```
[the remark this answers is not in the document]
Comment (smilesun) [reply examples/yaml/slurm/config.yaml:15]: @rahulbshrestha , could you add some comments here to each configuration?

**Review by smilesun — CHANGES_REQUESTED** [review]

**Comment by smilesun** [comment]
Error in rule run_experiment:
    output: zoutput/benchmarks/test_bloodcell_benchmark/rule_results/5.csv
[Wed Mar  8 12:58:29 2023]
Error in rule run_experiment:
    jobid: 6
    input: zoutput/benchmarks/test_bloodcell_benchmark/hyperparameters.csv
    output: zoutput/benchmarks/test_bloodcell_benchmark/rule_results/2.csv

[Wed Mar  8 12:58:29 2023]
rule run_experiment:
    input: zoutput/benchmarks/test_bloodcell_benchmark/hyperparameters.csv
    output: zoutput/benchmarks/test_bloodcell_benchmark/rule_results/2.csv
    jobid: 6
    reason: Missing output files: zoutput/benchmarks/test_bloodcell_benchmark/rule_results/2.csv; Input files updated by another job: zoutput/benchmarks/test_bloodcell_benchmark/hyperparameters.csv
    wildcards: index=2
    resources: mem_mb=60000, mem_mib=57221, disk_mb=1000, disk_mib=954,

File: examples/yaml/test_denbi_errors.txt [hunk 2]
```diff
@@ -0,0 +1,64 @@
+[Fri Feb 17 11:51:04 2023]
+Error in rule run_experiment:
```
[the remark this answers is not in the document]
Comment (rahulbshrestha) [reply examples/yaml/test_denbi_errors.txt]: @smilesun I think you're viewing an older version of this PR. This file has already been deleted.

File: domainlab/utils/hyperparameter_sampling.py [hunk 3]
```diff
@@ -160,6 +160,57 @@ def get_hyperparameter(name: str, config: dict) -> Hyperparameter:
     return SampledHyperparameter(name, config)
 
 
+class ReferenceHyperparameter(Hyperparameter):
```
[the remark this answers is not in the document]
Comment (rahulbshrestha) [reply domainlab/utils/hyperparameter_sampling.py]: @smilesun This file is not a part of this PR.

File: examples/benchmark/run_snakemake_slurm_conf_seed.sh [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+CONFIGFILE=$1
+export DOMAINLAB_CUDA_START_SEED=$2
+logdir="zoutput/logs"
+mkdir -p $logdir
+logfile="$logdir/$(timestamp).out"
+echo "Configuration file: $CONFIGFILE"
+echo "starting seed is: $DOMAINLAB_CUDA_START_SEED"
+echo "verbose log: $logfile"
+# Helmholtz
+snakemake --profile "examples/yaml/slurm" --keep-going --keep-incomplete --notemp --cores 5 -s "domainlab/exp_protocol/benchmark.smk" --configfile "$CONFIGFILE" 2>&1 | tee "$logfile"
```
Comment (smilesun) [inline examples/benchmark/run_snakemake_slurm_conf_seed.sh:18]: what does 2>&1 mean? could you comment there

**Review by smilesun — CHANGES_REQUESTED** [review]

**Comment by smilesun** [comment]
```
snakemake: error: Couldn't parse config file: while parsing a block mapping
  in "examples/yaml/slurm/config.yaml", line 2, column 1
expected <block end>, but found '<scalar>'
  in "examples/yaml/slurm/config.yaml", line 6, column 5

```

**Comment by smilesun** [comment]
@rahulbshrestha , 
for the current commit, if i run

`./examples/benchmark/run_snakemake_slurm_conf_seed.sh examples/yaml/test_helm_benchmark.yaml 0
`

I got 
```

Error executing rule run_experiment on cluster (jobid: 7, external: Submitted batch job 11333094, jobscript: /home/aih/xudong.sun/DomainLab/.snakemake/tmp.8lqmko_1/snakejob.run_experiment.7.sh). For error details see the cluster log and the log files of the involved rule(s).                                        
Trying to restart job 7.
Select jobs to execute...

```

there is no other useful information in the logs folder, does the same command work for you? 
which command did you use?

**Comment by smilesun** [comment]
@xinyuejohn @rahulbshrestha 
I ADDED some error logging in the configuration,
now i got the error

```
uleException:
FileNotFoundError in file /home/aih/xudong.sun/DomainLab/domainlab/exp_protocol/benchmark.smk, line 93:
[Errno 2] No such file or directory: '/lustre/groups/labs/marr/qscd01/datasets/armingruber/                            _Domains/Matek_cropped'

```

**Comment by smilesun** [comment]
i have re-introduced the style problems to check if everything could run.

**Comment by rahulbshrestha** [comment]
@smilesun hmm, this PR seems to have not have all the changes from the other PR we were working on. I'll fix it with John and let you know. The paths of the dataset need to be changed to the path we have in the helmholtz cluster.

**Comment by smilesun** [comment]
after fixing the path int the last commit, i got error which i can not diagnose from the --error output file, which read

```
cat run_experiment-index\=5-11335086.err 
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 2
Rules claiming more threads will be scaled down.
Provided resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954
Select jobs to execute...

[Wed Mar  8 19:12:32 2023]
rule run_experiment:
    input: zoutput/benchmarks/test_bloodcell_benchmark/hyperparameters.csv
    output: zoutput/benchmarks/test_bloodcell_benchmark/rule_results/5.csv
    jobid: 0
    reason: Missing output files: zoutput/benchmarks/test_bloodcell_benchmark/rule_results/5.csv
    wildcards: index=5
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, partition=gpu_p, qos=gpu

/home/aih/xudong.sun/anaconda3/lib/python3.9/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/home/aih/xudong.sun/anaconda3/lib/python3.9/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=AlexNet_Weights.IMAGENET1K_V1`. You can also use `weights=AlexNet_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)
/home/aih/xudong.sun/anaconda3/lib/python3.9/site-packages/torch/nn/_reduction.py:42: UserWarning: size_average and reduce args will be deprecated, please use reduction='sum' instead.
  warnings.warn(warning.format(ret))
/home/aih/xudong.sun/anaconda3/lib/python3.9/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: Metric `AUROC` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
/home/aih/xudong.sun/anaconda3/lib/python3.9/site-packages/torchmetrics/utilities/prints.py:36: UserWarning: No positive samples in targets, true positive value should be meaningless. Returning zero tensor in true positive score
  warnings.warn(*args, **kwargs)
```
I am interested to know what else is need to make those run.

**Comment by smilesun** [comment]
/var/spool/slurmd/job11335014/slurm_script: line 3: 1371902 Killed                  DOMAINLAB_CUDA_START_SEED='0' /home/aih/xudong.sun/anaconda3/bin/python -m snakemake --snakefile '/home/aih/xudong.sun/DomainLab/domainlab/exp_protocol/benchmark.smk' --target-jobs 'run_experiment:index=3' --allowed-rules 'run_experiment' --cores 'all' --attempt 2 --force-use-threads --resources 'mem_mb=1000' 'mem_mib=954' 'disk_mb=1000' 'disk_mib=954' --wait-for-files '/home/aih/xudong.sun/DomainLab/.snakemake/tmp.di7gu163' 'zoutput/benchmarks/test_bloodcell_benchmark/hyperparameters.csv' --force --keep-target-files --keep-remote --max-inventory-time 0 --nocolor --notemp --no-hooks --nolock --ignore-incomplete --keep-incomplete --rerun-triggers 'params' 'mtime' 'software-env' 'code' 'input' --skip-script-cleanup --use-conda --conda-frontend 'mamba' --conda-base-path '/home/aih/xudong.sun/anaconda3' --wrapper-prefix 'https://github.com/snakemake/snakemake-wrappers/raw/' --configfiles '/home/aih/xudong.sun/DomainLab/examples/yaml/test_helm_benchmark.yaml' --printshellcmds --latency-wait 60 --scheduler 'greedy' --scheduler-solver-path '/home/aih/xudong.sun/anaconda3/bin' --default-resources 'mem_mb=1000' 'disk_mb=max(2*input.size_mb, 1000)' 'tmpdir=system_tmpdir' 'partition=gpu_p' 'qos=gpu' --mode 2
slurmstepd: error: Detected 1 oom-kill event(s) in StepId=11335014.batch. Some of your processes may have been killed by the cgroup out-of-memory handler.

**Comment by xinyuejohn** [comment]
@smilesun @rahulbshrestha I am currently testing this on Helmholtz cluster. Will pin you again once I make it run

**Comment by smilesun** [comment]
the proces was killed by out-of-memory signal
